module/zfs/vdev.c

```diff
@@ -1157,22 +1156,20 @@ vdev_open_children(vdev_t *vd)
 	 * in a single thread so that the same thread holds the
 	 * spa_namespace_lock
 	 */
-	if (vdev_uses_zvols(vd)) {
+	if (vdev_uses_zvols(vd) || taskq_member(system_taskq, curthread)) {
 		for (c = 0; c < children; c++) {
 			vd->vdev_child[c]->vdev_open_error =
 			    vdev_open(vd->vdev_child[c]);
 			vd->vdev_nonrot &= vd->vdev_child[c]->vdev_nonrot;
 		}
 		return;
 	}
-	tq = taskq_create("vdev_open", children, minclsyspri,
-	    children, children, TASKQ_PREPOPULATE);
 
 	for (c = 0; c < children; c++)
-		VERIFY(taskq_dispatch(tq, vdev_open_child, vd->vdev_child[c],
-		    TQ_SLEEP) != 0);
+		VERIFY(taskq_dispatch(system_taskq, vdev_open_child,
+		    vd->vdev_child[c], TQ_SLEEP) != 0);
 
-	taskq_destroy(tq);
+	taskq_wait_outstanding(system_taskq, 0);
```

Comment:
This would wait for all tasks in the system_taskq to complete, won't it?
Don't we want to wait only for those which were dispatched in this function?
Imagine that a slow task is put on system_taskq and stays there for a long time. That would freeze vdev_open_children for no good reason.